src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/EndAppealHandler.java

```diff
@@ -63,9 +63,14 @@ public PreSubmitCallbackResponse<AsylumCase> handle(
                 .getCaseData();
 
         PaymentStatus paymentStatus = asylumCase.read(PAYMENT_STATUS, PaymentStatus.class)
-            .orElse(PaymentStatus.PAYMENT_PENDING);
-        if (callback.getEvent() == Event.END_APPEAL_AUTOMATICALLY && paymentStatus == PaymentStatus.PAID) {
-            throw new IllegalStateException("Cannot auto end appeal as the payment is already made!");
+                .orElse(PaymentStatus.PAYMENT_PENDING);
+
+        if (callback.getEvent() == Event.END_APPEAL_AUTOMATICALLY) {
+            if (paymentStatus == PaymentStatus.PAID) {
+                throw new IllegalStateException("Cannot auto end appeal as the payment is already made!");
+            } else if (callback.getCaseDetails().getState() == State.ENDED) {
+                throw new IllegalStateException("Appeal has already been ended!");
+            }
```

Comment:
I'm a fan of avoiding nesting when possible.

I have a suggestion for you:

```suggestion
            .orElse(PaymentStatus.PAYMENT_PENDING);
        if (callback.getEvent() == Event.END_APPEAL_AUTOMATICALLY && paymentStatus == PaymentStatus.PAID) {
            throw new IllegalStateException("Cannot auto end appeal as the payment is already made!");
        }

        if (callback.getEvent() == Event.END_APPEAL_AUTOMATICALLY && callback.getCaseDetails().getState() == State.ENDED) {
            throw new IllegalStateException("Appeal has already been ended!");
        }

```

Be advised that as long as the retry logic is as broken as is now, this will result in yet another exception being retried forever and ever. I'm still in the process of setting up the fix for the retry logic, which hopefully should be completed soon.